### PR TITLE
nixpkgs manual: advise against overriding whole phases

### PR DESCRIPTION
**Pull request by vcunat**
I've seen that mistake at least a few times already, e.g.
https://github.com/NixOS/nixpkgs/pull/26209#issuecomment-305925562
It might perhaps seem counter-intuitive if one doesn't know nixpkgs well.

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [x] Tested compilation of all (= none) pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/` = none)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by dezgeg** [comment]
I think the problem with Qt5 would be rather avoided by adding `_qtRmTmp` and `_qtFixCMakePaths` to  `$preFixupPhases` rather than `$postInstallHooks`.

**Comment by vcunat** [comment]
Right, that certainly seems less risky in this respect.  I guess @ttuegel knows best this stuff.

**Comment by ttuegel** [comment]
> I think the problem with Qt5 would be rather avoided by adding _qtRmTmp and _qtFixCMakePaths to $preFixupPhases rather than $postInstallHooks.

Then you have the same problem whenever someone overrides `fixupPhase`. That's less common, but it imposes the same constraint on packagers: don't override any whole phase without running the pre- and post-hooks.

**Comment by dezgeg** [comment]
No, changing `preFixupPhases` adds a new phase between `installPhase` and `fixupPhase`, whether they're overwritten or not (it's different from a `*Hook`).

Yes, you now get screwed if a package specifies `phases = ...` but that typically causes borkage in any package (because everyone who does that doesn't know they need `fixupPhase` so now you get manpages in wrong places or huge closures due to not shrinking rpaths).

**Comment by layus** [comment]
May I suggest to change the default builder in this respect.
For example, when "make install" does not work, using `installPhase = ...;` is much more natural than `dontInstall = true; postInstall = ...`. Plus, in the later case, you cannot be sure that your postInstall runse before postInstall hooks.

tl;dr; I think hooks should not be part of the phase, and should not be explicitly overriden if need be.

File: doc/stdenv.xml [hunk 1]
```diff
@@ -318,7 +318,13 @@ containing some shell commands to be executed, or by redefining the
 shell function
 <varname><replaceable>name</replaceable>Phase</varname>.  The former
 is convenient to override a phase from the derivation, while the
-latter is convenient from a build script.</para>
+latter is convenient from a build script.
+
+However, typically one only wants to <emphasis>add</emphasis> some
```
Comment (layus) [inline doc/stdenv.xml]: Well, yes, but what if I really want to _replace_ the default phase behavior, but keep the hooks. We should really explain that.

**Comment by gazally** [comment]
doc/stdenv.xml also contains [this](https://github.com/vcunat/nixpkgs/blob/8c6fada372dba386c25e209a4dff440e2da5a0be/doc/stdenv.xml#L66) example:
```
stdenv.mkDerivation {
  name = "fnord-4.5";
  ...
  buildPhase = ''
    gcc foo.c -o foo
  '';
  installPhase = ''
    mkdir -p $out/bin
    cp foo $out/bin
  '';
}
```
If redefining `installPhase` is not the right thing to do, then this example should be changed. However (as a newcomer to nixpkgs) I prefer @layus's solution. Keep the simple example for the simple package simple.

**Comment by layus** [comment]
Something weird happened. Apparently I did not post this message:

I have a working update to setup.sh that implements my above proposition in https://github.com/layus/nixpkgs/commit/66a2bdeb922a9ade5a7f1b031f362965e7ecc8ef

To test on only one package, you will need overrideSetup from pkgs/stdenv/adapters.nix.
In particular, it works for #26209.

**Comment by ttuegel** [comment]
> No, changing preFixupPhases adds a new phase between installPhase and fixupPhase, whether they're overwritten or not (it's different from a *Hook).

@dezgeg Oh, I understand now. Yes, I think that would be the most appropriate solution for the Qt 5 setup hooks. And I agree that although this may break packages that set `phases = ...`, generally one should not do that unless one understands very very well what one is doing!

**Comment by grahamc** [comment]
Seems pretty good, reminds me of this very unfortunate bit in the nixpkgs manual (a similar bit is in the nix manual) too:

```
While the standard environment provides a generic builder, you can still supply your own build script:

stdenv.mkDerivation {
  name = "libfoo-1.2.3";
  ...
  builder = ./builder.sh;
}
```
